pkg/parser/io_parser.go

```diff
@@ -0,0 +1,152 @@
+	return &IOParser{
+		inputBuffer:  make([]byte, 0, bufSize),
+		outputBuffer: make([]byte, 0, bufSize),
+		parse:        NewBufferedParser(parserBufSize, debug),
+	}
+}
+
+func (p *IOParser) Read(b []byte) (int, error) {
+	p.outputLock.Lock()
+	defer p.outputLock.Unlock()
+
+	var (
+		bufcap    = cap(b)
+		outbuflen = len(p.outputBuffer)
+	)
+
+	// If somehow we have exactly cap(b) in output buffer, just copy and return.
+	if outbuflen == bufcap {
+		copy(b, p.outputBuffer)
+		p.outputBuffer = p.outputBuffer[:0]
+		return bufcap, nil
+	}
+
+	// If we have more than cap(b), copy, resize and return.
+	if outbuflen > bufcap {
+		copy(b, p.outputBuffer[:bufcap])
+
+		// Proper way to resize the slice without losing it's length
+		copy(p.outputBuffer, p.outputBuffer[bufcap:])
+		p.outputBuffer = p.outputBuffer[:outbuflen-bufcap]
+
+		return bufcap, nil
+	}
+
+	// Not enough data left in output buffer, so drain it unconditionally.
+	copy(b[:outbuflen], p.outputBuffer)
+	p.outputBuffer = p.outputBuffer[:0]
+
+	// Remember how much we can write more
+	written := outbuflen
+
+	// Loop until we have either exhausted the parser OR the buffer.
+	for {
+		res, more := <-p.parse.Results()
+		if !more {
+			return written, io.EOF
+		}
+
+		data, err := json.Marshal(res)
+		if err != nil {
+			return written, err
+		}
+
+		var (
+			datalen = len(data)
+			capleft = bufcap - written
+		)
+
+		// We have exactly enough data to fill in the rest of the buffer, so copy and return.
+		if datalen == capleft {
+			copy(b[written:], data)
+			return bufcap, nil
+		}
+
+		// We have more data to write than we can fit in the output.
+		// Copy however many bytes we can from serialized result and store the rest in output buffer.
+		if datalen > capleft {
+			copy(b[written:], data[:capleft])
+			p.outputBuffer = append(p.outputBuffer, data[capleft:]...)
+			return bufcap, nil
+		}
+
+		// We can fit more than we have available for writing.
+		// Write as much as we can and continue to the next iteration.
+		copy(b[written:], data)
+		written = written + datalen
+	}
+}
+
+func (p *IOParser) Write(b []byte) (int, error) {
+	p.inputLock.Lock()
+	defer p.inputLock.Unlock()
+
+	written := 0
+
+	for {
+		buflen := len(b)
+
+		// Quick check for empty buffer.
+		if buflen == 0 {
+			return written, nil
+		}
+
+		// Check if incoming buffer contains a newline,
+		// since it indicates that we can now send the utterance to AST parser.
+		linesep := bytes.IndexRune(b, newline)
+
+		// If incoming bytestring doesn't contain a newline,
+		// then just append it to our "current line" buffer and return.
+		if linesep == -1 {
+			p.inputBuffer = append(p.inputBuffer, b...)
+			return written + buflen, nil
+		}
+
+		// Split the buffer by the newline character.
+		// Head contains the bytes [0 : newlineIndex]
+		// Re-slice the buffer to make sure we are progressively moving forward on it.
+		head := b[:linesep]
+		b = b[linesep+newlineLen:]
```

Comment:
What does this do?  `b` is not used again. And without returning `b` the slice doesn't really change in the caller. And if it did, it would break [io.Writer](https://golang.org/pkg/io/#Writer) specification that slice data must not change (well data data wouldn't change, just the slice indices).

Reply:
> What does this do?

As the comment states: "Re-slice the buffer to make sure we are progressively moving forward on it.". In other words, this will change `b` into a subset of `b`, starting at `linesep+newlineLen`, e.g. if `b = []byte{ 0, 1, 2, 3, 4}` and offset is `2`, after re-slicing, `b = []byte{2, 3, 4}`.

> b is not used again.

It is used, in the next iteration of the for-loop. This is done so that if `b` contains multiple "lines" (byte sequences, delimited by newline), in this loop we progressively move forward on `b` by searching and reading the next line in each iteration of the loop.